### PR TITLE
Include Graylog CA into places where we connect to the DataNode

### PR DESCRIPTION
**Pull request by janheise**
## Description

Adding the CA into several places.

/nocl

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java [hunk 1]
```diff
@@ -50,6 +57,36 @@ public static X509Certificate sign(PrivateKey caPrivateKey, X509Certificate caCe
                 Date.from(validFrom), Date.from(validUntil),
                 csr.getSubject(), csr.getSubjectPublicKeyInfo());
 
+
+        // see if we have the dns/rfc822/ip address extensions specified in the csr
+
+        ArrayList<GeneralName> altNames = new ArrayList<>();
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java:63]: This one could be initialized insider the if-statement

File: graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java [hunk 2]
```diff
@@ -50,6 +57,36 @@ public static X509Certificate sign(PrivateKey caPrivateKey, X509Certificate caCe
...
+                GeneralName[] names = gns.getNames();
+                for (int i = 0; i < names.length; i++) {
+                    switch (names[i].getTagNo()) {
+                        case GeneralName.dNSName:
+                        case GeneralName.iPAddress:
+                        case GeneralName.rfc822Name:
+                            altNames.add(names[i]);
+                            break;
+                    }
+                }
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java:81]: Won't it look nicer if replaced by streams and lambdas?

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightResource.java [hunk 3]
```diff
@@ -79,10 +79,19 @@ public List<DataNode> listDataNodes() {
 
         return activeDataNodes.values().stream().map(n -> {
             final var preflight = preflightDataNodes.get(n.getNodeId());
-            return new DataNode(n.getNodeId(), n.getType(), n.getTransportAddress(), preflight != null ? preflight.state() : null, preflight != null ? preflight.errorMsg() : null, n.getHostname(), n.getShortNodeId());
+            return new DataNode(n.getNodeId(), n.getType(), removeCredentialsFromTransportAddress(n.getTransportAddress()), preflight != null ? preflight.state() : null, preflight != null ? preflight.errorMsg() : null, n.getHostname(), n.getShortNodeId());
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightResource.java:82]: Aren't there more places where you have to hide credentials?
(like ClusterResource? )

Comment (janheise) [reply graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightResource.java:82]: I made a note in our project board so we fix this in a separate PR

File: graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java [hunk 4]
```diff
@@ -29,11 +35,20 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Date;
 
 import static org.graylog.security.certutil.CertConstants.SIGNING_ALGORITHM;
 
 public class CsrSigner {
+    private static boolean isValidName(final int name) {
+        return switch (name) {
+            case GeneralName.dNSName, GeneralName.iPAddress, GeneralName.rfc822Name -> true;
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java:47]: What about a static import to make this line shorter?

File: graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java [hunk 5]
```diff
@@ -50,6 +65,24 @@ public static X509Certificate sign(PrivateKey caPrivateKey, X509Certificate caCe
...
+        if (certAttributes != null && certAttributes.length > 0) {
+            ArrayList<GeneralName> altNames = new ArrayList<>();
+
+            for (Attribute attribute : certAttributes) {
+                Extensions extensions = Extensions.getInstance(attribute.getAttrValues().getObjectAt(0));
+                GeneralNames gns = GeneralNames.fromExtensions(extensions, Extension.subjectAlternativeName);
+                if (gns == null) {
+                    continue;
+                }
+                Arrays.stream(gns.getNames()).filter(n -> isValidName(n.getTagNo())).forEach(altNames::add);
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog/security/certutil/csr/CsrSigner.java:78]: What about removing `continue` ?
```
                if (gns != null) {
                     Arrays.stream(gns.getNames()).filter(n -> isValidName(n.getTagNo())).forEach(altNames::add);
                }
```

File: graylog2-server/src/main/java/org/graylog2/security/CustomCAX509TrustManager.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class CustomCAX509TrustManager implements X509TrustManager {
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/security/CustomCAX509TrustManager.java:37]: Could this class be unit-tested?

**Review by luk-kaminski — COMMENTED** [review]
I have found one potential error when testing.
After the start-up I can access data-nodes OS with the HTTPS url : https://lkaminski-graylog:9200
But surprisingly, I can also use HTTP url: http://localhost:8999/ and get the following response:
```{
  "opensearch": {
    "opensearch_version": "2.4.1",
    "node": {
      "node_name": "node1",
      "state": "AVAILABLE",
      "is_leader_node": true,
      "rest_base_url": "https://admin:changeit@lkaminski-graylog:9200",
      "process": {
        "pid": 15478,
        "alive": true,
        "started": "2023-07-19T09:45:04.740Z"
      }
    }
  },
  "data_node_version": "5.2.0-SNAPSHOT"
}```

I just wonder - shouldn't insecure access be blocked completely after data-node is configured ?

**Review by luk-kaminski — APPROVED** [review]
Looks fine.
I have left a few comments, but I am ok if they are solved in a separate PRs.

File: graylog2-server/src/main/java/org/graylog2/security/CustomCAX509TrustManager.java (hunk 6 above)
Comment (janheise) [reply graylog2-server/src/main/java/org/graylog2/security/CustomCAX509TrustManager.java:37]: added a test

**Comment by todvora** [comment]
@luk-kaminski regarding the mix of https for opensearch and http for the datanode itself - these are two completely independent cases, but we could also reuse the ssl cert to secure the datanode interface. But I'd suggest to solve this in a different PR.

**Review by todvora — APPROVED** [review]
LGTM

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightResource.java (hunk 3 above)
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightResource.java:82]: Will be solved in #16055

**Pull request merged**